### ISSUE TITLE
misdefined test method kills trial with an unhandled exception

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1901]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1901|
|Type|defect|
|Created|2006-07-06 12:24:53Z|

Like so:

```
exarkun@charm:~$ cat test_foo.py
from twisted.trial.unittest import TestCase

class Foo(TestCase):
    def test_foo():
        pass
exarkun@charm:~$ trial test_foo.py 
Running 1 tests.
test_foo
  Foo
    test_foo ...                                                        [ERROR]

===============================================================================
[ERROR]: test_foo.Foo.test_foo

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/bin/trial", line 24, in ?
    run()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/scripts/trial.py", line 377, in run
    test_result = trialRunner.run(suite)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/runner.py", line 503, in run
    result.printErrors()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/reporter.py", line 277, in printErrors
    self._formatFailureTraceback)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/reporter.py", line 255, in _printResults
    self.write(formatter(*(content[1:])))
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/reporter.py", line 245, in _formatFailureTraceback
    fail.frames, frames = self._trimFrames(fail.frames), fail.frames
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/reporter.py", line 235, in _trimFrames
    last = newFrames[-1]
IndexError: list index out of range
exarkun@charm:~$ 
```

<details><summary>Searchable metadata</summary>

```
trac-id__1901 1901
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1152188693000000 1152188693000000
changetime__1172434897000000 1172434897000000
version__None None
owner__ 
cc__therve cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Hit this again while writing tests for loopbackAsync.  Here's the code that did it:

```python

from twisted.trial.unittest import TestCase
from twisted.internet.protocol import Protocol
from twisted.internet.defer import Deferred
from twisted.internet.interfaces import IAddress
from twisted.protocols.loopback import loopbackAsync

class LoopbackAsyncTestCase(TestCase):
    def _hostpeertest(self, get, testServer):
        """
        Test one of the permutations of client/server host/peer.
        """
        class TestProtocol(Protocol):
            def makeConnection(self, transport):
                self.connected.callback(transport)

        if testServer:
            server = TestProtocol()
            d = server.connected = Deferred()
            client = Protocol()
        else:
            server = Protocol()
            client = TestProtocol()
            d = client.connected = Deferred()

        loopbackAsync(server, client)

        def connected(transport):
            host = getattr(transport, get)()
            self.failUnless(IAddress.providedBy(host))

        return d.addCallback(connected)


    def test_serverHost(self, get):
        """
        Test that the server gets a transport with a properly functioning
        implementation of L{ITransport.getHost}.
        """
        return self._hostpeertest("getHost", True)
```

Not sure if this is caused by the same bug or not.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Ah, nevermind.  Of course it is the same case.  I accidentally left the extra argument on test_serverHost.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

Ready to review in test-arguments-1901.

I modified _trimFrames and add tests to the problem. The best as exarkun said would be to remove _trimFrames, but let's live with this for now.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

The twisted.internet.defer import within the definition of the ContainMalformed class seems unnecessary.  Also please add a docstring for that class describing how it is useful for the tests which use it.

Grammatically, the TestMalformedMethod docstring should read:

```
Test that trial manages when test methods don't have good signatures.
```

"trial" and "manages" need to agree; since trial is signular, manages gets the "s".  If trial were replaced with a plural noun (eg, "test runners") then "manage" would be correct.

Likewise, "methods" and "do" (as it appears in the contraction "don't") need to agree for the same reason.

(I hope that kind of feedback is useful, if not, just let me know and I'll skip it next time :)

As a non-grammatical comment, I would suggest "correct" in the place of "good".

test_extraArg has a typo of "has" spelled "as".

In reporter.py, the added comment:

```
There is also another case, when the test method is bad defined and contains extra argument
```

Should read

```
There is also another case, when the test method is badly defined and contains extra arguments.
```

"badly" is the adverb form of the adjective "bad", which is required since it is modifying "defined", which is a verb, not a noun.

Otherwise the change seems reasonable.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve removed owner|
|-|-|

Replying to [exarkun](#note_6):
> The twisted.internet.defer import within the definition of the ContainMalformed class seems unnecessary.

It's used for deferredGenerator, because it changes the signature of the method.

>  Also please add a docstring for that class describing how it is useful for the tests which use it.

OK, done.

> (I hope that kind of feedback is useful, if not, just let me know and I'll skip it next time :)

It is, I'm not as careful as I'd want to be :). I've update the docstrings according to your comments, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve|
|-|-|

Oops, sorry.  With respect to the defer import, I meant that it was unnecessary for it to be inside the class definition, instead of at the top of the module. :)  Sorry for being unclear.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

Sorry for being dumb :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

Looks good to merge now. :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [19657]) Merge test-arguments-1901

Author: therve
Reviewer: exarkun
Fixes [#1901](https://github.com/twisted/twisted/issues/1901)

Fix a problem with misdefined test methods within trial: report the good error
message and don't kill trial.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `reopened`|
|-|-|

For some reason, this introduced 8 failures on Windows. :(

See r19662 or buildbot for details.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve removed owner<br>@therve set status to `new`|
|-|-|

That should be good now, for some reasons it needed a specific directory to operator.

For record, TestUntilFailure is dangerous because if you introduce a second test in the method it begin to fail the same way.

I try to launch a build on win32 slave.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

The 2 builds on win32-py-2.4 and win32-py-2.5 has succeeded.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @therve|
|-|-|

Some stuff I guess I missed last time:

  * os import at the top of test_runner.py should come before z.i import.  in general, stdlib imports should come first, then third party imports, then twisted imports.
  * I think it's good to give each test its own directory, but I think the fact that this is /required/ for the tests to pass indicates a bug in trial.  In particular, it seems as though _tearDownLogging isn't being called at all the right times.  Perhaps this merits another ticket?

Aside from those issues (and the 2nd one isn't even a blocker for this), looks good to merge, so please do after the import order is fixed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set status to `closed`|
|-|-|

(In [19711]) Merge test-arguments-1901

Author: therve
Reviewer: exarkun
Fixes [#1901](https://github.com/twisted/twisted/issues/1901)

Fix a problem with misdefined test methods within trial: report the good error
message and don't kill trial.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
